### PR TITLE
[depends: #1339] Get goto model

### PR DESCRIPTION
**Pull request by kroening**
This is a new and simpler way to arrive at a goto program.

### CODE REVIEW

File: src/goto-programs/goto_convert_functions.h [hunk 1]
```diff
@@ -25,7 +25,6 @@ void goto_convert(
 
 // confusing, will go away
```
Comment (tautschnig) [inline src/goto-programs/goto_convert_functions.h]: Is this comment still true?

File: src/goto-programs/get_goto_model.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  {
+    goto_modelt tmp_model;
+
+    if(is_goto_binary(filename))
+    {
+      message.status() << "Reading GOTO program from file "
+                       << messaget::eom;
+
+      if(read_goto_binary(filename, tmp_model, message_handler))
+        throw 0;
```
Comment (tautschnig) [inline src/goto-programs/get_goto_model.cpp]: How is the actual error going to be conveyed?

File: src/goto-programs/get_goto_model.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include "goto_convert_functions.h"
+
+static symbol_tablet get_source(
+  const cmdlinet &cmdline,
+  const std::string &filename,
+  message_handlert &message_handler)
+{
+  messaget message(message_handler);
+
+  languaget *language=get_language_from_filename(filename);
```
Comment (tautschnig) [inline src/goto-programs/get_goto_model.cpp]: How is this _not_ leaking memory?

File: src/goto-programs/get_goto_model.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  #ifdef _MSC_VER
+  std::ifstream infile(widen(filename));
+  #else
+  std::ifstream infile(filename);
+  #endif
+
+  if(!infile)
+  {
+    message.error() << "failed to open input file `"
```
Comment (tautschnig) [inline src/goto-programs/get_goto_model.cpp]: Shouldn't this also do
```
   source_locationt location;
   location.set_file(filename);
   message.error().source_location=location;
```
as is done above?

Comment (kroening) [inline src/goto-programs/get_goto_model.cpp]: Contemplated it, but then we'd be sending out filenames that we know don't exist.

File: src/goto-programs/get_goto_model.cpp (hunk 2 above)
Comment (kroening) [inline src/goto-programs/get_goto_model.cpp]: Message handler!

File: src/cbmc/cbmc_parse_options.cpp [hunk 5]
```diff
@@ -478,39 +480,84 @@ int cbmc_parse_optionst::doit()
...
-      ui_message_handler);
+      get_message_handler());
 
   // get solver
-  cbmc_solverst cbmc_solvers(options, symbol_table, ui_message_handler);
-  cbmc_solvers.set_ui(get_ui());
+  cbmc_solverst cbmc_solvers(
+    options,
+    goto_model.symbol_table,
+    ui_message_handler);
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: `get_message_handler()` ?

File: src/cbmc/cbmc_parse_options.cpp [hunk 6]
```diff
@@ -527,21 +574,25 @@ int cbmc_parse_optionst::doit()
 
   prop_convt &prop_conv=cbmc_solver->prop_conv();
 
-  bmct bmc(options, symbol_table, ui_message_handler, prop_conv);
+  bmct bmc(
+    options,
+    goto_model.symbol_table,
+    ui_message_handler,
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: `get_message_handler()` ?

File: src/cbmc/cbmc_parse_options.cpp [hunk 7]
```diff
@@ -788,49 +738,44 @@ void cbmc_parse_optionst::preprocessing()
...
-    namespacet ns(symbol_table);
-
     // Remove inline assembler; this needs to happen before
     // adding the library.
-    remove_asm(symbol_table, goto_functions);
+    remove_asm(goto_model);
 
     // add the library
-    link_to_library(symbol_table, goto_functions, ui_message_handler);
+    link_to_library(goto_model, ui_message_handler);
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: `get_message_handler()` ?

File: src/cbmc/cbmc_parse_options.cpp [hunk 8]
```diff
@@ -843,67 +788,66 @@ bool cbmc_parse_optionst::process_goto_program(
...
     // add loop ids
-    goto_functions.compute_loop_numbers();
+    goto_model.goto_functions.compute_loop_numbers();
 
     if(cmdline.isset("drop-unused-functions"))
     {
       // Entry point will have been set before and function pointers removed
       status() << "Removing unused functions" << eom;
-      remove_unused_functions(goto_functions, ui_message_handler);
+      remove_unused_functions(goto_model, ui_message_handler);
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: `get_message_handler()` ?

**Comment by tautschnig** [comment]
The failing CI tests indicate that there is work left to be done.

**Comment by kroening** [comment]
#1339 prepares this.

**Comment by thomasspriggs** [comment]
I am going to close this PR as it appears that work on it has been inactive for an extended period of time. This is not any judgement on how worthwhile this work is. This is part of an effort to reduce the number of open PRs which are not being actively worked on. If you still think it would be worthwhile to get this merged then the branch should be re-based on the latest version of develop and the PR re-opened.

**Pull request closed without merging**